Add support for virt domain tags

mojaves opened the pull request:
After the discussion in https://github.com/collectd/collectd/pull/2037 , here's the same code implemented on top of the existing codebase.
As explained in https://github.com/collectd/collectd/pull/2037 , we are considering to switch to collectd in oVirt for monitoring hosts and libvirt domains in future releases.

We can't just neglect to monitor all the domains if one fraction thereof is unresponsive, so we need to enhance the monitoring code to work with that.

We will implement that on top of the more generic domain tag/partition approach I implemented here.
This patch required some refactoring, but I tried to keep this minimal.

I choose to upload few patches to make the review easier, but I have zero problems squashing them.

More code could be backported from https://github.com/collectd/collectd/pull/2037 - tests and documentations, and it will be done with later patches or later PRs.

rubenk commented on src/virt.c:
```diff
@@ -135,8 +164,8 @@ static enum plginst_field plugin_instance_format[PLGINST_MAX_FIELDS] =
-	target,
-	source
+        target,
+        source
```
If you're fixing the indentation, might as well make it 2 spaces like the other enums :)

mojaves replied:
done in a followup patch, could be squashed if you wish

rubenk commented on src/virt.c:
```diff
@@ -480,16 +500,41 @@ lv_config (const char *key, const char *value)
 {
     time_t t;
+    struct lv_read_instance *inst = ud->data;
+    struct lv_read_state *state = &inst->read_state;
```
Move this down below the check on the follow line?

mojaves replied:
right, fixed

rubenk commented on src/virt.c:
```diff
@@ -102,11 +92,50 @@ struct interface_device {
-static int add_interface_device (virDomainPtr dom, const char *path, const char *address, unsigned int number);
+#define NR_INSTANCES_DEFAULT 1
+#define NR_INSTANCES_MAX 128
+static size_t nr_instances = NR_INSTANCES_DEFAULT;
```
size_t is a bit large for something with a max of 128. Just make this an int?

mojaves replied:
why not, I'm checking boundaries anyway in lv_config.

rubenk commented:
Hi @mojaves!

Thanks, I think this looks good, I just left some minor suggestions inline.
I think at minimum this needs some documentation for the tag format.

If you don't mind I'll wait with merging this until after we branch 5.7, which should be in a week or two.

mojaves commented:
@rubenk sure, there is no rush. I'll address your comments, backport docs and tests (in followup patches) and do more testing myself.

mojaves commented:
@rubenk 
Let me document my last updates:

- Addressed your comments - with followup patches, hopefully this is easier to review; feel free to squash if you prefer, either way is fine for me

- Backported the tests , but disabled them (see https://github.com/collectd/collectd/pull/2048/commits/f246c0df8de83a1635bbdecd8f88a29d1763ea53 to see the reason)

- Added documentation (https://github.com/collectd/collectd/pull/2048/commits/b3eb9e778149be7cacb7e1bfd375a8f33c2790b7)

- And finally manage to squash one bug which exists even in the old code (https://github.com/collectd/collectd/pull/2048/commits/32ad52204ef1cfa28eee8501a443d312c20e327e)

I've done testing on my side - both with backward-compatible and with the new tag support enable with no issues.

mojaves commented:
@rubenk @octo  I guess I need to rebase after the reformatting, right?

rubenk commented:
@mojaves yes please, and if you could also pull your changes through clang-format that would be great.

mojaves commented:
@rubenk rebase done. Unfortunately I had to force push, and this messed up the order of commits.
Here's the expected order:

999efc2 virt plugin: Document the partition/tag support
0936a9e virt plugin: Properly reset the lists
6188d7d virt plugin: Disable tests momentarily
98e38ad virt plugin: Add tests for the domain tags
19828ad virt plugin: Minor fixes
a726d25 virt plugin: Use int, for the number of instances
c84646c virt plugin: Avoid NULL pointer deference
452324c virt plugin: Add support for domain tags
9f9af17 virt plugin: Support for multiple read callbacks
2846042 virt plugin: Factor the read state into a struct

Please also note that "virt plugin: Fix bad rebase of virt_test" was squashed into "virt plugin: Add tests for the domain tags" and that "virt plugin: Fix the indentation of enums" was dropped because is no longer relevant

rubenk commented:
Thanks @mojaves!

rubenk commented:
Hi @mojaves, I was just looking at the error handling in lv_domain_get_tag, specifically https://github.com/collectd/collectd/blob/master/src/virt.c#L794.
On error, lv_domain_get_tag jumps to the done label, but we always return the result of xmlXPathRegisterNs().

Also, the err = 0 assignment on https://github.com/collectd/collectd/blob/master/src/virt.c#L779 is a no-op.

 Would you mind sending a follow up PR that fixes this?

Thanks!

mojaves commented:
@rubenk sure thing: https://github.com/collectd/collectd/pull/2093
